docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+    // KeyFromString returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    KeyFromString(buf *Buffer, str string) error
+
+    /******* Possible Optimization ********/
+    // FromKey returns the original byte slice from collation key.
+    FromKey(buf *Buffer, str []byte) error
+    // StringFromKey returns the original string from collation key.
+    StringFromKey(str []byte) string
+}
+```
+The interface is quite similar to the Go [collate package](https://godoc.org/golang.org/x/text/collate), it is possible to implement some of the MySQL collation by simply mapping it to an existing Go collation.
+
+### Row Format
+
+The encoding layout of TiDB has been described in our [previous article](https://pingcap.com/blog/2017-07-11-tidbinternal2/#map). The row format should be changed to make it memory comparable, this is important to the index lookup. Basic principle is that all keys encoded should use the `sortKeys` result from `Key()`/`KeyFromString()` function. However, most of the `sortKeys` calculations are not reversible.
+
+#### Option 1
+
+  * For table data, encodings stay unchanged. All strings are compared after decoding with the `Compare()` function.
+  * For table indices, we replace current `ColumnValue` with `sortKey`:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+     Since index value can not be recovered from `sortKey`, extra table lookup is needed for current Index-only queries, this may lead to performance regressions. The extra table lookup also requires corresponding changes to the TiDB optimizer.
+
+#### Option 2
+
+  * Table data encodings is the same with Option 1.
+  * For table indices, we encode the `ColumnValue` to the value of the storage, so as to avoid extra table lookup for index queries:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID_indexedColumnsValue
+	```
+    - For non-unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: indexedColumnsValue
+	```
+
+Pros: Extra table lookup is avoided and no change to the TIDB optimizer.
+
+Cons: The size of index value on string column with new collations is doubled(we need to write both `sortKey` and `indexedColumnsValue`, their sizes are equal for most of the collations).
+
+## Compatibility
+
+### Compatibility with MySQL
+
+As stated before, MySQL has various different collation implementations. For example, in MySQL 5.7, the default collation of `utf8mb4` is `utf8mb4_general_ci`, which has two main flaws:
+  * It only implement the `simple mappings` of UCA.
+  * It used a very old version of Unicode.
```

Comment:
`utf8mb4_general_ci` is the default, but `utf8mb4_unicode_520_ci` is already available and more correct. See: https://mysqlserverteam.com/sushi-beer-an-introduction-of-utf8-support-in-mysql-8-0/

I think the most important limitation is that it only supports accent weight. So if you want case sensitivity, you need to use a binary collation, which is crazy.. because then you lose sort order.

Reply:
> I think the most important limitation is that it only supports accent weight.

Do you mean that we cannot have a case-sensitive but accent-insensitive collation?

Reply:
Sorry, I meant that it only supports case insensitive (character, accent, punctuation) collations. So sensitivity requires a switch to binary. I've edited this to be correct.

Reply:
Now MySQL 8 supports case sensitive collations:

```
mysql> show collation where charset='utf8mb4';                                                            
+----------------------------+---------+-----+---------+----------+---------+---------------+
| Collation                  | Charset | Id  | Default | Compiled | Sortlen | Pad_attribute |
+----------------------------+---------+-----+---------+----------+---------+---------------+
| utf8mb4_0900_ai_ci         | utf8mb4 | 255 | Yes     | Yes      | 0       | NO PAD        |
| utf8mb4_0900_as_ci         | utf8mb4 | 305 |         | Yes      | 0       | NO PAD        |
| utf8mb4_0900_as_cs         | utf8mb4 | 278 |         | Yes      | 0       | NO PAD        |
| utf8mb4_0900_bin           | utf8mb4 | 309 |         | Yes      | 1       | NO PAD        |
| utf8mb4_bin                | utf8mb4 | 46  |         | Yes      | 1       | PAD SPACE     |
...
```

Since the full UCA weight-level comparison is implemented. TiDB can do this as well.